File: modules/remote-check/bukkit/src/main/java/ru/privatenull/pnlibrary/remote/bukkit/RemoteCheckContext.java
package ru.privatenull.pnlibrary.remote.bukkit;

import org.bukkit.plugin.java.JavaPlugin;
import ru.privatenull.pnlibrary.common.minecraft.MinecraftVersion;
import ru.privatenull.pnlibrary.common.minecraft.MinecraftVersionInfo;

import java.util.Collections;
import java.util.Map;

/** Safe, immutable data exposed to a remote policy class. */
public final class RemoteCheckContext {
    private final JavaPlugin plugin;
    private final String pluginVersion;
    private final String serverVersion;
    private final MinecraftVersion minecraftVersion;
    private final MinecraftVersionInfo minecraftVersionInfo;
    private final Map<String, String> values;

    RemoteCheckContext(JavaPlugin plugin, Map<String, String> values) {
        this.plugin = plugin;
        this.pluginVersion = plugin.getDescription().getVersion();
        this.serverVersion = plugin.getServer().getVersion();
        this.minecraftVersionInfo = MinecraftVersion.parseInfo(serverVersion);
        this.minecraftVersion = minecraftVersionInfo.getParsed();
        this.values = Collections.unmodifiableMap(values);
    }

    public JavaPlugin plugin() { return plugin; }
    public String pluginVersion() { return pluginVersion; }
    public String serverVersion() { return serverVersion; }
    /** Parsed Minecraft version; UNKNOWN is returned when the server exposes an unknown release. */
    public MinecraftVersion minecraftVersion() { return minecraftVersion; }
    /** Full parse result; retains the raw server version when the enum is UNKNOWN. */
    public MinecraftVersionInfo minecraftVersionInfo() { return minecraftVersionInfo; }
    public Map<String, String> values() { return values; }
}

File: platforms/bukkit/runtime/src/main/kotlin/ru/privatenull/pnlibrary/remote/bukkit/RemoteCheckContext.kt
package ru.privatenull.pnlibrary.remote.bukkit

import org.bukkit.plugin.java.JavaPlugin
import ru.privatenull.pnlibrary.common.minecraft.MinecraftVersion
import ru.privatenull.pnlibrary.common.minecraft.MinecraftVersionInfo
import java.util.Collections
import java.util.Locale

class RemoteCheckContext internal constructor(
    val plugin: JavaPlugin,
    values: Map<String, String>,
) {
    val pluginVersion: String = plugin.description.version
    val serverVersion: String = plugin.server.version
    val minecraftVersionInfo: MinecraftVersionInfo = MinecraftVersion.parseInfo(serverVersion)
    val minecraftVersion: MinecraftVersion = minecraftVersionInfo.parsed
    val serverInfo: BukkitServerInfo = BukkitServerInfo(
        detectPlatform(plugin.server.name), plugin.server.name, serverVersion, minecraftVersionInfo,
    )
    val values: Map<String, String> = Collections.unmodifiableMap(HashMap(values))

    fun plugin() = plugin
    fun pluginVersion() = pluginVersion
    fun serverVersion() = serverVersion
    fun minecraftVersion() = minecraftVersion
    fun minecraftVersionInfo() = minecraftVersionInfo
    fun serverInfo() = serverInfo
    fun values() = values

    private fun detectPlatform(name: String): BukkitPlatform {
        val normalized = name.lowercase(Locale.ROOT)
        return when {
            "purpur" in normalized -> BukkitPlatform.PURPUR
            "folia" in normalized -> BukkitPlatform.FOLIA
            "paper" in normalized -> BukkitPlatform.PAPER
            "spigot" in normalized -> BukkitPlatform.SPIGOT
            "craftbukkit" in normalized -> BukkitPlatform.CRAFTBUKKIT
            "bukkit" in normalized -> BukkitPlatform.BUKKIT
            else -> BukkitPlatform.UNKNOWN
        }
    }
}
